### ISSUE TITLE
Runtime DI

### ISSUE THREAD

**Issue by jwaterloo** [boost-ext/di#116]
How does one go about grafting a runtime di library onto yours? Coming from the Java world when it comes to DI, the ideal is Spring for runtime and Dagger for compile time. Yours would be Dagger but runtime DI ie. configuration is an important part of the mix. Think runtime graph, possibly boost serialization, being dependent on DI compile time graph and vice versa. ... Is such a thing possible? Could be called "zipper". Doesn't have to be xml could be property file/map nested_name to value.

**Comment by kris-jusiak** [comment]
Yep, proposed boost.di is more like dagger2 (although it compiles faster! and has way better runtime performance - examples here https://github.com/krzysztof-jusiak/di_frameworks).

Anyway, moving on to runtime support, there is a support for it in the library:

```
auto runtime_flag = true;
auto injector = di::make_injector(
     di::bind<interface>.to([&](const auto& injector) {
          return runtime_flag ? 
                injector.create<shared_ptr<implementation1>>() :
                injector.create<shared_ptr<implementation2>>()
     }
);

injector.create<shared_ptr<interface>>() // implementation1
runtime_flag = false;
injector.create<shared_ptr<interface>>() // implementation2
```

https://github.com/krzysztof-jusiak/di/blob/cpp14/example/dynamic_bindings.cpp

If it comes to configuration from xml, it's possible as well:
https://github.com/krzysztof-jusiak/di/blob/cpp14/example/extensions/injections/xml_injection.cpp

If you could elaborate about what you mean by "zipper" I will definitely try to implement it using DI. There is a possibility to serialize using the library, but it won't work with members not being injected by constructor, unless you are saying about dumping types using DI from the composition root and use boost serialization to serialize each of them, that's defo possible and pretty easy to write and actually a nice idea :) 

P.S. I have integrated DI with Boost.MSM already where DI is responsible for creating actions/guards - https://github.com/krzysztof-jusiak/msm

Thanks for your question, it's really useful.

**Comment by kris-jusiak** [comment]
close it for now as there is no response in 18 days

**Comment by jwaterloo** [comment]
zipper is a name I would give to a library if dagger2 and Spring decided to make a baby; left of zipper is compile time, right of zipper run time ie. spring like xml/properties configuration file or vice versa, both sides of the zipper is a graph and would require zig zagging between compile time and runtime to make sure everything was hooked up in the right order

**Comment by kris-jusiak** [comment]
okay, I get what you mean and agree that making sure that everything is hooked up properly would be really important in this case

well, some time ago I was trying to merge both, run-time and compile-time, worlds but I gave up as I figured out how to do run-time stuff with compile-time approach as well as handling exceptions and compile time errors together was really time consuming

anyway, in the link below you can find my initial approach
https://github.com/krzysztof-jusiak/di_runtime_injector

if it comes to compile-time approach, besides run-time binding, you can also use type erasure injector on top of it

```
di::injector<i> injector = di::make_injector(
    di::bind<i>.to<impl>()
);

injector.create<std::unique_ptr<i>>();
```

**Comment by jwaterloo** [comment]
Please don't get hung up on this one. Just a wishlist and possibly impossible going back and forth anyway. Just future ideas. Dagger2 like compile time only with minimal runtime examples, documentation and getting this into Boost is most important. The C++ community tends to be compile time crazy anyway. That having been said, throw the following on your [distant] radar.
1) factory methods/objects [This may just be a documentation issue as not much different from a constructor.]
2) [static] factory functions [This may just be a documentation issue as not much different from a constructor.]
...
3) tooling; this seems lacking in other DI but you already have nice example of generating plantUML, highlight this in your documentation, explain how you created your pretty pictures, upsale your strengths, how writing compile time results in good documentation for review. There may be other tools in the future that brings value to developers and thus makes your DI shine. Frankly, I look at Spring DI as more of a configuration framework and as such still prefer to use their xml and original properties file configuration over their compile time configuration. CDI/Juice didn't make any sense at first glance, more work, high costs and little value. It was only when I learned a little dagger[2] and Boost.DI did dependency injection made any sense. That's because it was explained in terms of a compile time graph and the developer work to put and maintain things in the right order.
...
40) property based DI; compile time property literals [This may be needed by 50)]
https://dzone.com/articles/typesafe-property-literals
https://github.com/Doctusoft/ds-bean
...
50) configuration [files]; this may take care of the minimal runtime requirements
a) database connection dependency coming from configuration file or LDAP
b) other objects on compile time graph being loaded from database
NOTE: Here, I am talking about setting properties after the constructors have been called, ie. Hooks. The trick here is being given a ?JSON? path from a common entry point. This may be where the module as in dagger2 comes into the picture. Each module being its own compile time graph.
...
160) runtime DI; an additional step that occurs totally after compile time, runtime time can depend upon compile time but compile time can not depend upon runtime; this could be better than spring if it allows changing configuring the graph at runtime and those changes persisted

**Comment by kris-jusiak** [comment]
Thank you for your suggestions, they are very, very useful and will totally use them!

Anyway, I'm not getting the property stuff yet, but I will defo look into it and response with more adequate info.

If it comes to run-time di, well I do get your points, but IMHO all of it might be done with boost.di approach right now as well.

For example XML injection:
http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/extensions/xml_injection.html

To sum up, I agree that the first step has to be to get it into boost, but I'm really exited to hear how many things might be achieved with dependency injection in the future!
